### PR TITLE
Refactor the way we create routes

### PR DESCRIPTION
**Pull request by hibikir**
The current way of creating routes requires creating some complex types by hand, and uses implicits as a way to ascertain the legality of the selected options. I think that's just overkill, and it puts quite a bit of an onus on the users to understand the internal types.

Instead, I am using a sum type, and abusing the already existing implicates to chain with each other. It has its own disadvantages, but IMO it's a lot less than the original design.

A warning: This IS a breaking change for users, but the fix is typically to use the delete key to remove unnecessary code.

### CODE REVIEW

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala [hunk 1]
```diff
@@ -167,22 +168,22 @@ object `AWS::EC2::Route` extends DefaultJsonProtocol {
     def read(json: JsValue) = ???
```
Comment (kevinmeredith) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: I know that you didn't write this, but perhaps you know why there's a `???` here?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala [hunk 2]
```diff
@@ -109,14 +109,15 @@ object `AWS::EC2::CustomerGateway` extends DefaultJsonProtocol {
...
-  implicit object valid2 extends ValidRouteCombo[None.type , Some[ResourceRef[`AWS::EC2::Instance`]], None.type]
-  implicit object valid3T extends ValidRouteCombo[None.type , None.type, Some[Token[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]]]
-  implicit object valid3 extends ValidRouteCombo[None.type , None.type, Some[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]]
+sealed trait ValidRouteComboOption
+case class InternetGatewayRoute(v:Token[ResourceRef[`AWS::EC2::InternetGateway`]]) extends ValidRouteComboOption
+case class EC2InstanceRoute(v:Token[ResourceRef[`AWS::EC2::Instance`]]) extends ValidRouteComboOption
+case class VPCPeeringRoute(v:Token[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]) extends ValidRouteComboOption
+
+object ValidRouteComboOption {
+  implicit def toInternetGateway[T](v: T)(implicit t:T => Token[ResourceRef[`AWS::EC2::InternetGateway`]]) = InternetGatewayRoute(v)
```
Comment (kevinmeredith) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: Can you replace the `T` and `implicit` with:

```
implicit def toInternetGateway(v: Token[ResourceRef[`AWS::EC2::InternetGateway`]) 
```
?

Comment (hibikir) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: Implicit view resolution is only recursive if the implicit view itself needs another implicit. So I do it for chaining. In this case, we need chaining: We turn a Resource into a Token[ResourceRef] with an implicit, and we want to combine them both for free.

http://docs.scala-lang.org/tutorials/FAQ/chaining-implicits.html

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala (hunk 1 above)
Comment (hibikir) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: Yes, it's more implicit chinning: Some code really wants DefaultJsonProtocols, even though all we ever want to provide is DefaultJonWriters

**Comment by tj-corrigan** [comment]
looks like travis hiccuped getting dependencies, tested clean locally

**Pull request merged**